In [1]:
# Install dependencies on jnbk
#!pip install -r requirements.txt

In [2]:
## Import modules
import pandas as pd
import seaborn as sns
import matplotlib as mpl
import matplotlib.pyplot as plt
#import pandas_bokeh
import os
import sys

In [4]:
## Check modules version
print('Notebook running from:',sys.executable)
print('Pandas:',pd.__version__)
print('Matplotlib:',mpl.__version__)
print('Seaborn:',sns.__version__)

Notebook running from: /home/naneen/Documents/DataScience/Projects/FinHack_LT_AV/py3-finhack/bin/python3
Pandas: 1.2.1
Matplotlib: 3.3.4
Seaborn: 0.11.1


In [5]:
## List data-raw
path_pwd = os.getcwd()
path_data_train = path_pwd+'/'+'data_raw/Train/' 
os.listdir(path_data_train)

['train_bureau.xlsx', 'data_dict.xlsx', '.DS_Store', 'train_Data.xlsx']

In [6]:
## Set trainig data files
file_customer_train = path_data_train+'train_Data.xlsx'
file_bureau_train   = path_data_train+'train_bureau.xlsx'
print('Customer file:', file_customer_train)
print('Bureau file:', file_bureau_train)

Customer file: /home/naneen/Documents/DataScience/Projects/FinHack_LT_AV/data_raw/Train/train_Data.xlsx
Bureau file: /home/naneen/Documents/DataScience/Projects/FinHack_LT_AV/data_raw/Train/train_bureau.xlsx


In [13]:
## Load customer trainig data
df_customer = pd.read_excel(file_customer_train)
df_customer.head()

ID Frequency InstlmentMode LoanStatus   PaymentMode  BranchID      Area  \
0   1   Monthly        Arrear     Closed         PDC_E         1       NaN   
1   2   Monthly       Advance     Closed           PDC       333    BHOPAL   
2   3   Quatrly        Arrear     Active  Direct Debit         1       NaN   
3   7   Monthly       Advance     Closed        Billed       125      GUNA   
4   8   Monthly        Arrear     Closed        Billed       152  BILASPUR   

   Tenure  AssetCost  AmountFinance  ...  ManufacturerID  SupplierID    LTV  \
0      48     450000       275000.0  ...            1568       21946  61.11   
1      47     485000       350000.0  ...            1062       34802  70.00   
2      68     690000       519728.0  ...            1060      127335  69.77   
3      48     480000       400000.0  ...            1060       25094  80.92   
4      44     619265       440000.0  ...            1046       21853  71.05   

  SEX   AGE  MonthlyIncome         City           State   ZiPCODE  \
0   M  49.0       35833.33       RAISEN  MADHYA PRADESH  464993.0   
1   M  23.0         666.67       SEHORE  MADHYA PRADESH  466001.0   
2   M  39.0       45257.00       BHOPAL  MADHYA PRADESH  462030.0   
3   M  24.0       20833.33  ASHOK NAGAR  MADHYA PRADESH  473335.0   
4   M  56.0       27313.67     BILASPUR     CHATTISGARH  495442.0   

        Top-up Month  
0        > 48 Months  
1  No Top-up Service  
2       12-18 Months  
3        > 48 Months  
4       36-48 Months  

[5 rows x 26 columns]

In [16]:
## Size of customer training data
df_customer.shape

(128655, 26)

**Obesrvations**:
- Loading customer _xlsx_ data has been successful.

***

### Exploratory Data Analysis on Customer data

In [50]:
# Run the python script to load util functions
# similar to import modules functionality
%run './utils.py'

In [41]:
# Loop over customer columns list to display datatypes of attribute
show_features_datatypes(df_customer)

Column id:   0 	Name: ID           	DataType: int64
Column id:   1 	Name: Frequency    	DataType: object
Column id:   2 	Name: InstlmentMode 	DataType: object
Column id:   3 	Name: LoanStatus   	DataType: object
Column id:   4 	Name: PaymentMode  	DataType: object
Column id:   5 	Name: BranchID     	DataType: int64
Column id:   6 	Name: Area         	DataType: object
Column id:   7 	Name: Tenure       	DataType: int64
Column id:   8 	Name: AssetCost    	DataType: int64
Column id:   9 	Name: AmountFinance 	DataType: float64
Column id:  10 	Name: DisbursalAmount 	DataType: float64
Column id:  11 	Name: EMI          	DataType: float64
Column id:  12 	Name: DisbursalDate 	DataType: datetime64[ns]
Column id:  13 	Name: MaturityDAte 	DataType: datetime64[ns]
Column id:  14 	Name: AuthDate     	DataType: datetime64[ns]
Column id:  15 	Name: AssetID      	DataType: int64
Column id:  16 	Name: ManufacturerID 	DataType: int64
Column id:  17 	Name: SupplierID   	DataType: int64
Column id:  18 	Na

### **Observations:**
- Issues that could/should be addressed during feature engineering are listed here. 
- **Frequency** is object datatype. Can be easily changed to a numerical value. 
- **InstlmentMode** can be an imporatant attribute that we can change to numerical value. 
- **LoanStatus** is a simple yes/no. So, we could change into binary values 1/0 i.e., One-hot encoding! 
- **PaymentMode** as explained in data dict file needs some processing.
- **Sex** could be changed into numerical variable.
- **Top-up Month** is the target variable that we will be trying to predict. We should definitely look into this attribute as an object datatyep for this attributre makes no sense to me. Let's not forget about this!

***

In [45]:
# Drop duplicate rows from dataframe, if any!
drop_duplicate_rows(df_customer)

There are 0 duplicated rows in the dataset.


In [44]:
## Customer colnames into a list
list_customer_columns = df_customer.columns.to_list()

In [51]:
## Loop over list of customer columns to get summary
for icnt, colname in enumerate(list_customer_columns):
    print('Column ID:',icnt)
    show_feature_summary(df_customer, colname, display_uniques=False)

Column ID: 0
Details of feature: ID
         - datatype: int64
         - col.size: (128655,)
         - NaN.vals: 0


Column ID: 1
Details of feature: Frequency
         - datatype: object
         - col.size: (128655,)
         - NaN.vals: 0


Column ID: 2
Details of feature: InstlmentMode
         - datatype: object
         - col.size: (128655,)
         - NaN.vals: 0


Column ID: 3
Details of feature: LoanStatus
         - datatype: object
         - col.size: (128655,)
         - NaN.vals: 0


Column ID: 4
Details of feature: PaymentMode
         - datatype: object
         - col.size: (128655,)
         - NaN.vals: 0


Column ID: 5
Details of feature: BranchID
         - datatype: int64
         - col.size: (128655,)
         - NaN.vals: 0


Column ID: 6
Details of feature: Area
         - datatype: object
         - col.size: (128655,)
         - NaN.vals: 11653


Column ID: 7
Details of feature: Tenure
         - datatype: int64
         - col.size: (128655,)
         - NaN.va

### **Observations:**:
- There are missing values that should be addressed. 
-  Missing values are found in:
    - **Area**: 11653
    - **MaturityDAte**: 1
    - **SEX**: 59
    - **AGE**: 59
    - **MonthlyIncome**: 234
    - **City**: 11256
    - **ZiPCODE**: 372

***

### Target attribute:
**Top-up Month** is the target variable to predict. Let's see how it looks. 

In [59]:
df_customer['Top-up Month'].value_counts()

No Top-up Service    106677
 > 48 Months           8366
36-48 Months           3656
24-30 Months           3492
30-36 Months           3062
18-24 Months           2368
12-18 Months           1034
Name: Top-up Month, dtype: int64

***

## Features/Attributes 

### Frequency:

In [69]:
df_customer['Frequency'].value_counts()

Half Yearly    76248
Monthly        31150
Quatrly        20795
BI-Monthly       462
Name: Frequency, dtype: int64

### InstllmentMode

In [70]:
df_customer['InstlmentMode'].value_counts()

Arrear     122349
Advance      6306
Name: InstlmentMode, dtype: int64

### LoanStatus

In [71]:
df_customer['LoanStatus'].value_counts()

Closed    94457
Active    34198
Name: LoanStatus, dtype: int64

### PaymentMode

In [72]:
df_customer['PaymentMode'].value_counts()

Direct Debit    31766
ECS             31390
PDC             26617
Billed          26486
PDC_E            9937
Auto Debit        843
SI Reject         744
Cheque            442
ECS Reject        417
Escrow              7
PDC Reject          6
Name: PaymentMode, dtype: int64

### BranchID

In [73]:
df_customer['BranchID'].value_counts()

24     4474
199    4194
50     2842
166    2381
202    2208
       ... 
139       8
400       5
408       4
395       2
386       2
Name: BranchID, Length: 189, dtype: int64

***

### Phase-1 PyCaret